RecCalorimeter/src/components/CreatePositionedCaloCells.cpp

```diff
@@ -0,0 +1,189 @@
+
+// k4geo
+#include "detectorCommon/DetUtils_k4geo.h"
+
+// edm4hep
+#include "edm4hep/CalorimeterHit.h"
+
+DECLARE_COMPONENT(CreatePositionedCaloCells)
+
+CreatePositionedCaloCells::CreatePositionedCaloCells(const std::string& name, ISvcLocator* svcLoc) :
+Gaudi::Algorithm(name, svcLoc) {
+  declareProperty("hits", m_hits, "Hits from which to create cells (input)");
+  declareProperty("cells", m_cells, "The created calorimeter cells (output)");
+
+  declareProperty("positionsTool", m_cellPositionsTool, "Handle for cell positions tool");
+  declareProperty("crosstalkTool", m_crosstalkTool, "Handle for the cell crosstalk tool");
+  declareProperty("calibTool", m_calibTool, "Handle for tool to calibrate Geant4 energy to EM scale tool");
+  declareProperty("noiseTool", m_noiseTool, "Handle for the calorimeter cells noise tool");
+  declareProperty("geometryTool", m_geoTool, "Handle for the geometry tool");
+}
+
+StatusCode CreatePositionedCaloCells::initialize() {
+  StatusCode sc = Gaudi::Algorithm::initialize();
+  if (sc.isFailure()) return sc;
+
+  info() << "CreatePositionedCaloCells initialized" << endmsg;
+  info() << "do calibration : " << m_doCellCalibration << endmsg;
+  info() << "add cell noise      : " << m_addCellNoise << endmsg;
+  info() << "remove cells below threshold : " << m_filterCellNoise << endmsg;
+  info() << "emulate crosstalk : " << m_addCrosstalk << endmsg;
+
+  // Initialization of tools
+
+  // Cell position tool
+  if (!m_cellPositionsTool.retrieve()) {
+    error() << "Unable to retrieve the cell positions tool!!!" << endmsg;
+    return StatusCode::FAILURE;
+  }
+  // Cell crosstalk tool
+  if (m_addCrosstalk) {
+    if (!m_crosstalkTool.retrieve()) {
+      error() << "Unable to retrieve the cell crosstalk tool!!!" << endmsg;
+      return StatusCode::FAILURE;
+    }
+  }
+  // Calibrate Geant4 energy to EM scale tool
+  if (m_doCellCalibration) {
+    if (!m_calibTool.retrieve()) {
+      error() << "Unable to retrieve the calo cells calibration tool!!!" << endmsg;
+      return StatusCode::FAILURE;
+    }
+  }
+  // Cell noise tool
+  if (m_addCellNoise || m_filterCellNoise) {
+    if (!m_noiseTool.retrieve()) {
+      error() << "Unable to retrieve the calo cells noise tool!!!" << endmsg;
+      return StatusCode::FAILURE;
+    }
+    // Geometry settings
+    if (!m_geoTool.retrieve()) {
+      error() << "Unable to retrieve the geometry tool!!!" << endmsg;
+      return StatusCode::FAILURE;
+    }
+    // Prepare map of all existing cells in calorimeter to add noise to all
+    StatusCode sc_prepareCells = m_geoTool->prepareEmptyCells(m_cellsMap);
+    if (sc_prepareCells.isFailure()) {
+      error() << "Unable to create empty cells!" << endmsg;
+      return StatusCode::FAILURE;
+    }
+  }
+
+  // Copy over the CellIDEncoding string from the input collection to the output collection
+  auto hitsEncoding = m_hitsCellIDEncoding.get_optional();
+  if (!hitsEncoding.has_value()) {
```

Comment:
Is this worthy of crashing, can't this be just a warning?

Reply:
it's exactly as in CreateCaloCells (https://github.com/HEP-FCC/k4RecCalorimeter/blob/02d796b43562ce1c4188d7ef95258a53366a1745/RecCalorimeter/src/components/CreateCaloCells.cpp#L78). Without encoding the input cells can't be properly decoded in terms of layer and theta/eta/phi id which is needed for the various tools (positioning, calibration, ..).
I